emmeans and all pairwise comparisons (#43)
beacon-biosignals/Effects.jl: docs/src/emmeans.md

The "marginal" here refers to estimation at the margins of the table, i.e. either averaging over or computing at specific values of other variables, often the means of other variables.
Effects.jl provides a convenient interface for computing EM means at all levels of categorical variables and the means of any continuous covariates. An example will make this clear:

```@example emmeans
using DataFrames, Effects, GLM, StatsModels, StableRNGs, StandardizedPredictors
rng = StableRNG(42)
growthdata = DataFrame(; age=[13:20; 13:20],
                       sex=repeat(["male", "female"], inner=8),
                       weight=[range(100, 155; length=8); range(100, 125; length=8)] .+ randn(rng, 16))
model_scaled = lm(@formula(weight ~ 1 + sex * age), growthdata;
                  contrasts=Dict(:age => ZScore(), :sex => DummyCoding()))

emmeans(model_scaled)
```

(This is the same example data as [Interaction Terms in Effects](@ref).)

Notice that we could have achieved the same result by using `effects` and specifying the levels manually:

```@example emmeans
design = Dict(:sex => ["female", "male"], :age => 16.5)
grid = expand_grid(design)
effects!(grid, model_scaled)
```

> `emmeans` is primarily a convenience function for computing calling effects at a convenient, pre-defined reference grid.

Reviewer: ```suggestion
`emmeans` is primarily a convenience function for computing effects on a convenient, pre-defined reference grid.
```

<!-- XXX
should add documentation about using Satterthwaite and Kenward-Roger but
there's not a Julia implementation of those things yet, so not high priority
-->

### Multiple Comparisons Correction

A perhaps more important concern is inflation of Type-1 error rate through multiple testing.
We can also provide a correction method to be applied to the p-values.
```@example emmeans
using MultipleTesting
bonferroni(pvals) = adjust(PValues(pvals), Bonferroni())
empairs(mixed_model; dof=Inf, padjust=bonferroni)
```

[^GLMMFAQ]: [Why doesn’t lme4 display denominator degrees of freedom/p values? What other options do I have?](https://bbolker.github.io/mixedmodels-misc/glmmFAQ.html#why-doesnt-lme4-display-denominator-degrees-of-freedomp-values-what-other-options-do-i-have)
[^Bates2006]: [lmer, p-values and all that](https://stat.ethz.ch/pipermail/r-help/2006-May/094765.html)

## A Final Note

There are many subleties in more advanced applications of EM means.
These are discussed at length in the documentation for the R package `emmeans`.
For example:

- [the impact of transformations and link functions](https://cran.r-project.org/web/packages/emmeans/vignettes/transformations.html)

> - [derived covariates](https://cran.r-project.org/web/packages/emmeans/vignettes/basics.html#depcovs),  e.g., including both `x` and `x^2` in a model

Reviewer: ```suggestion
- [derived covariates](https://cran.r-project.org/web/packages/emmeans/vignettes/basics.html#depcovs), e.g., including both `x` and `x^2` in a model
```
